solvation_analysis/analysis.py

```diff
@@ -0,0 +1,127 @@
+        #     self.kernel_kwargs = {}
+
+        self.solute = solute
+        self.solvents = solvents
+        self.u = self.solute.universe
+        self.rdf_plots = {}
+        self.rdf_data = {}
+        self.ion_speciation = None
+        self.ion_pairing = None
+        self.coordination_numbers = None
+        self.solvation_frames = []
+
+    @classmethod
+    def _plot_solvation_radius(cls, bins, data, radius):
+        fig, ax = plt.subplots()
+        ax.plot(bins, data, "b-", label="rdf")
+        ax.axvline(radius, color="r", label="solvation radius")
+        ax.set_xlabel("Radial Distance (A)")
+        ax.set_ylabel("Probability Density")
+        ax.legend()
+        return fig, ax
+
+    def run_prepare(self):
+        for name, solvent in self.solvents.items():
+            # generate and save RDFs
+            rdf = InterRDF(self.solute, solvent, range=(0.0, 8.0))
+            # TODO: specify start and stop args
+            rdf.run()
+            bins, data = rdf.results.bins, rdf.results.rdf
+            self.rdf_data[name] = (data, bins)
+            # generate and save plots
+            if name not in self.radii.keys():
+                self.radii[name] = self.kernel(bins, data, **self.kernel_kwargs)
+            fig, ax = self._plot_solvation_radius(bins, data, self.radii[name])
+            ax.set_title(f"Solvation distance of {name}")
+            self.rdf_plots[name] = fig, ax
+
+    def _prepare(self):
+        assert self.solvents.keys() == self.radii.keys(), "Radii missing."
+        # columns: solute #, atom id, distance, solvent name, res id
+
+    def _single_frame(self):
+        # initialize empty arrays
+        all_pairs = np.empty((0, 2), dtype=np.int)
+        all_dist = np.empty(0)
+        all_tags = np.empty(0)
+        for name, solvent in self.solvents.items():
+            pairs, dist = distances.capped_distance(
+                self.solute.positions,
+                solvent.positions,
+                self.radii[name],
+                box=self.u.dimensions,
+            )
+            # replace local ids with absolute ids
+            pairs[:, 1] = solvent.ids[[pairs[:, 1]]]  # TODO: ids vs ix?
+            # extend
+            all_pairs = np.concatenate((all_pairs, pairs))
+            all_dist = np.concatenate((all_dist, dist))
+            all_tags = np.concatenate((all_tags, np.full(len(dist), name)))
+        # put the data into a data frame
+        all_resid = self.u.atoms[all_pairs[:, 1]].resids
+        solvation_data_np = np.column_stack(
+            (all_pairs[:, 0], all_pairs[:, 1], all_dist, all_tags, all_resid)
+        )
+        solvation_data_pd = pd.DataFrame(
+            solvation_data_np,
+            columns=["solvated_atom", "atom_id", "dist", "res_name", "res_id"],
+        )
+        self.solvation_frames.append(solvation_data_pd)
+
+    def _conclude(self):
+        # OPTIONAL
+
+        self.ion_speciation = _IonSpeciation(self.solvation_frames)
```

Comment:
Also on the theme of my above comment, IMO these should take class data objects rather than a dataframe, then make the dataframe an optional step at the end.

That way also each method can take only the information it needs rather than a huge `df` with all the solvation data. :)

Reply:
This is a great idea. By "class data objects" do you mean a proper @dataclass? 

Perhaps what I will do is define a data class that is the solvation data and then have each of the methods access it. I agree that the current organization is not optimal.

Reply:
Sorry my terminology wasn't very clear. What I should have called them is `data members`.  I would go with `Solution` having a few attributes for the data eg `self.pairs`.

You can use a `@dataclass` if you would like, but I personally don't see the advantage in separating the data from the class used to generate it.

Reply:
Note though that all `@dataclass` does is auto generate `__init__` and `__repr__` and other than that is still just a regular class.

Reply:
Having a "block" of data would be convenient. I mainly worry about polluting the Solution class with a bunch of attributes that aren't meant to be accessed. If I created a data class it would let me split up the code where:
Solution: generates raw solvation data and accesses processed data through analysis methods
Dataclass: organizes raw data and calculates key attributes of data
Analysis classes: process data and provide methods for accessing processed data

I am unsure if it's worth having an extra class though.

Reply:
Plan seems reasonable to me. :)

What do @MDAnalysis/gsoc-mentors think?